### ISSUE TITLE
Scala 2 Regression: Upper bound of a higher-kinded type variable is not included into implicit search scope

### ISSUE THREAD

**Issue by neko-kai** [lampepfl/dotty#13986]
## Compiler version
 
`3.1.2-RC1-bin-20211118-2b37257-NIGHTLY`

## Minimized code

Scastie: https://scastie.scala-lang.org/mBqRVwrfQpqZeLBUKzf8xQ

```scala
package example

sealed trait Xa[T]
sealed trait Mu[T] extends Xa[T]
object Xa {
  implicit def convertMu[X[x] <: Xa[x], A, B](implicit t: X[A]): X[B] = t.asInstanceOf[X[B]]
}
object Mu {
  implicit def mu: Mu[Int] = new Mu[Int]
}

object App extends App {
  def constrain(a: Mu[Long]): Unit = println(a)
  constrain(Xa.convertMu)
}
```

## Output

```scala
no implicit argument of type X[A] was found for parameter t of method convertMu in object Xa

where:    A is a type variable
          X is a type variable with constraint <: example.Mu


The following import might fix the problem:

  import example.Mu.mu
```

## Expectation

Expected to work, as in Scala 2: https://scastie.scala-lang.org/MesjnzXeRCSarfLr7HCGHA
This seems to affect only higher-kinded types, I could not reproduce it with a non-parameterized implicit with a roughly similar setup - https://scastie.scala-lang.org/XW8Yy4CXQmSXgslU0Wktew

* NOTE 1: Specifying _just_ the type `Mu` explicitly, e.g. using partial application below will get rid of the error:
  ```scala
  def convertMuPartial[X[x] <: Xa[x]]: [A, B] => X[A] ?=> () => X[B] = {
    [A, B] => (t: X[A]) ?=> () => Xa.convertMu[X, A, B](t)
  }
  constrain(convertMuPartial[Mu]())
  ```
  The implicit scope gets lost only when the searchable becomes a type parameter at some point in the search.

* NOTE 2: The searched-for type is determined correctly and applying the suggestion to `import example.Mu.mu` will get rid of the error. However, this import should not be required because Mu companion object should be included in implicit scope, brought in by constraint on X <: example.Mu.

In addition, varying the above setup slightly produces an `assertion failure` compiler crash: \<a new issue link will be added here shortly\>

This error currently blocks me from porting Izumi BIO to Scala 3, where a similar implicit is used https://github.com/7mind/izumi/blob/develop/fundamentals/fundamentals-bio/src/main/scala/izumi/functional/bio/Root.scala#L120

**Comment by prolativ** [comment]
The snippet is missing `{}` after `new Mu[Int]`.

Besides that this seems to be a double regressing as adding the suggested import makes the code compile with 3.1.1-RC1 but not with 3.1.2-RC1-bin-20211118-2b37257-NIGHTLY

**Comment by neko-kai** [comment]
@prolativ Good catch! I didn't check if the import actually works on Nightly, as I started minimizing with 3.1.0 and only confirmed the negative result on Nightly.

**Comment by neko-kai** [comment]
Hi, this regression is still preventing the port of https://github.com/izumi/izumi to Scala 3
